# lesson 3:
    
data: cflo_v3.3.gff.gz

这是Camponotus floridanus蚂蚁的蛋白编码基因的注释文件，以Gff3格式存储。

任务是：1）网上查找资料了解Gff3格式；2）统计该蚂蚁的基因总个数；3）计算每一个基因的CDS的总长，输出格式为第一列是基因ID，第二列是该基因的CDS总长度。

## answer

GFF3是注释文件。文件中每一行分为9列，以TAB分开。

依次是：

1. reference sequence：参照序列

 指出注释的对象。如一个染色体，克隆或片段。可以有多个参照序列。

2. source：来源

 注释的来源。如果未知，则用点（.）代替。

3. type：类型

 属性的类型。如gene，repeat_region，exon，CDS等。

4. start position：起点

 属性对应片段的起点。从1开始计数。

5. end position：终点

 属性对应片段的终点。

6. score：得分

 对于一些可以量化的属性，可以在此设置一个数值以表示程度的不同。如果为空，用点（.）代替。

7. strand：链

 “＋”表示正链，“－”表示负链，“.”表示不需要指定正负链。

8. phase：步进

 对于编码蛋白质的CDS来说，本列指定下一个密码子开始的位置。可以是0，1或2，表示到达下一个密码子需要跳过的碱基个数。对于其它属性，则用点（.）代替。

9. attributes：属性

## 数据统计

In [1]:
# lesson3 practice
# date: 2018/08/01
# author: zxzhu

in_filename = './data/Traning/Lesson3/cflo_v3.3.gff'
out_filename = 'lesson3_result.txt'
gene = {}
f = open(in_filename, 'r')
for line in (i.strip() for i in f):
    if 'CDS' in line:
        line = line.split('\t')
        gene_name = line[-1].split('=')[-1][:-1]
        gene[gene_name] = gene.get(gene_name,0) + int(line[4]) - int(line[3] + 1)
        
f.close()
print('Genes:\t{}.\n'.format(len(gene.keys())))

with open(out_filename, 'w') as w:
    w.write('Gene_name\tCDS_length\n')
    w.write('\n'.join([key + '\t' + str(value) for key, value in gene.items()]))
    

Genes:	17064.



共计17064个基因，每个基因的CDS总长度见：[lesson3_result](https://github.com/zxzhu/BGI-Practice/blob/master/data/lesson3_result.txt)

## Reference

[GFF3文件格式](https://blog.csdn.net/hugolee123/article/details/44302231)